x/emissions/module/rewards/rewards.go

```diff
@@ -159,26 +176,49 @@ func getDistributionAndPayoutRewardsToTopicActors(
 
 func CalcTopicRewards(
 	ctx sdk.Context,
-	k keeper.Keeper,
 	weights map[uint64]*alloraMath.Dec,
 	sortedTopics []uint64,
-	sumWeight alloraMath.Dec,
-	totalReward alloraMath.Dec,
+	sumWeight alloraMath.Dec, // sum of all topic weights
+	totalReward alloraMath.Dec, // maxTotalReward in treasury
+	epochLengths map[uint64]int64, // epoch lengths for each topic
+	currentBlockEmissionDec alloraMath.Dec, // perBlockEmission
 ) (
 	map[uint64]*alloraMath.Dec,
 	error,
 ) {
+	totalTopicRewardsSum := alloraMath.ZeroDec()
 	topicRewards := make(map[TopicId]*alloraMath.Dec)
 	for _, topicId := range sortedTopics {
-		topicRewardFraction, err := GetTopicRewardFraction(weights[topicId], sumWeight)
+		topicWeight := weights[topicId]
+		topicRewardFraction, err := GetTopicRewardFraction(topicWeight, sumWeight)
 		if err != nil {
 			return nil, errors.Wrapf(err, "topic reward fraction error")
 		}
-		topicReward, err := GetTopicReward(topicRewardFraction, totalReward)
+		if alloraMath.ZeroDec().Equal(topicRewardFraction) {
+			ctx.Logger().Warn(fmt.Sprintf("Skipping rewards for topic: %d, zero weights", topicId))
+			continue
+		}
+		topicRewardPerBlock, err := GetTopicReward(topicRewardFraction, currentBlockEmissionDec)
 		if err != nil {
 			return nil, errors.Wrapf(err, "topic reward error")
 		}
-		topicRewards[topicId] = &topicReward
+		epochLength := epochLengths[topicId]
+		if epochLength <= 0 {
+			return nil, errors.Wrapf(types.ErrInvalidLengthTopic, "epoch length is nil or zero for topic %d", topicId)
+		}
+		epochLengthDec := alloraMath.NewDecFromInt64(epochLength)
+		topicRewardPerEpoch, err := topicRewardPerBlock.Mul(epochLengthDec)
```

Comment:
I'm not sure this is right / what @jmdkastro meant. I think this is nearly always going to trip the `totalTopicRewardsSum.Gt(totalReward)` condition below (henceforth: "the condition below").

Per [his comment](https://linear.app/alloralabs/issue/RES-424/asynchronous-topic-weightpayout-fix#comment-fc74dda0), I believe we have to normalize by the unit time "over which we want this [pot] to be depleted". This is the `[time]` term he mentions when he says `dt/[time]`.

Even with this division, I'm not convinced that we're going to reliably *not* trip the condition below. I'm wondering if we have to rebalance/renormalize rewards per topic upon satisfaction of that condition, and/or if we should apply some heuristic that guarantees (perhaps with some v high probability) the total rewards will be less than whatever is left in the rewards pot.

Maybe we need to define a "global normalization divisor" that gets updated upon every topic inactivation and activation.

Reply:
I did all the algebra and actually I see now: you're right, I'm wrong! The terms cancel out and yeah you only need to multiply by `[time]`. Very good.